File: Leetcode/Design/155_min-stack/main.cpp
// 155. 最小栈
// Created by kaide on 2021/5/28.
//
#include <stack>

using namespace std;

class MinStack {
public:
    /** initialize your data structure here. */
    MinStack() {
        minStk.push(INT_MAX);
    }

    void push(int val) {
        stk.push(val);
        if (val < minStk.top()) {
            minStk.push(val);
        } else {
            minStk.push(minStk.top());
        }
    }

    void pop() {
        stk.pop();
        minStk.pop();
    }

    int top() {
        return stk.top();
    }

    int getMin() {
        return minStk.top();
    }

private:
    stack<int> minStk;
    stack<int> stk;
};

int main() {
    MinStack stk = MinStack();
    stk.push(2);
    stk.push(0);
    stk.push(-3);
    printf("%d\n", stk.getMin());   // expect:-3
    stk.pop();
    printf("%d\n", stk.top());   // expect: 0
    printf("%d\n", stk.getMin());   // expect: 0
}

File: Leetcode/Design/155_min-stack/main.py
import sys


class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.min_stk, self.stk = [], []

    def push(self, x: int) -> None:
        self.stk.append(x)
        min = self.min_stk[-1]
        if min > x:
            self.min_stk.append(x)
        else:
            self.min_stk.append(min)

    def pop(self) -> None:
        self.stk.pop()
        self.min_stk.pop()

    def top(self) -> int:
        return self.stk[-1]

    def min(self) -> int:
        return self.min_stk[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.min()
if __name__ == '__main__':
    stk = MinStack()
    stk.push(2)
    stk.push(0)
    stk.push(-3)
    print(stk.min())  # expect:-3
    stk.pop()
    print(stk.top())  # expect: 0
    print(stk.min())  # expect: 0
